cleanrl/ddpg_continuous_action.py

```diff
@@ -148,95 +98,132 @@ def forward(self, x, a, device):
-        qf1_loss = loss_fn(qf1_a_values, next_q_value)
-
-        # optimize the midel
-        q_optimizer.zero_grad()
-        qf1_loss.backward()
-        nn.utils.clip_grad_norm_(list(qf1.parameters()), args.max_grad_norm)
-        q_optimizer.step()
-
-        if global_step % args.policy_frequency == 0:
-            actor_loss = -qf1.forward(s_obs, actor.forward(s_obs, device), device).mean()
-            actor_optimizer.zero_grad()
-            actor_loss.backward()
-            nn.utils.clip_grad_norm_(list(actor.parameters()), args.max_grad_norm)
-            actor_optimizer.step()
-
-            # update the target network
-            for param, target_param in zip(actor.parameters(), target_actor.parameters()):
-                target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
-            for param, target_param in zip(qf1.parameters(), qf1_target.parameters()):
-                target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
-
-        if global_step % 100 == 0:
-            writer.add_scalar("losses/qf1_loss", qf1_loss.item(), global_step)
-            writer.add_scalar("losses/actor_loss", actor_loss.item(), global_step)
-
-    # TRY NOT TO MODIFY: CRUCIAL step easy to overlook 
-    obs = next_obs
-
-    if done:
-        # TRY NOT TO MODIFY: record rewards for plotting purposes
-        print(f"global_step={global_step}, episode_reward={episode_reward}")
-        writer.add_scalar("charts/episodic_return", episode_reward, global_step)
-        obs, episode_reward = env.reset(), 0
 
-env.close()
-writer.close()
+if __name__ == "__main__":
+    args = parse_args()
+    run_name = f"{args.gym_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
+    if args.track:
+        import wandb
+
+        wandb.init(
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            sync_tensorboard=True,
+            config=vars(args),
+            name=run_name,
+            monitor_gym=True,
+            save_code=True,
+        )
+    writer = SummaryWriter(f"runs/{run_name}")
+    writer.add_text(
+        "hyperparameters",
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # TRY NOT TO MODIFY: seeding
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    device = torch.device("cuda" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = gym.vector.SyncVectorEnv(
+        [make_env(args.gym_id, 0, 0, args.capture_video, run_name)]
+    )
+    assert isinstance(envs.single_action_space, gym.spaces.Box), "only continuous action space is supported"
+
+    max_action = float(envs.single_action_space.high[0])
+    actor = Actor(envs).to(device)
+    qf1 = QNetwork(envs).to(device)
```

Comment:
Probably a bit nit-picky, but everywhere there is a need to create a copy, such as `qf1_target` for example, we could simply use `qf1_target = qf1.clone()`, making it slightly cleaner.

Additionally, disabling gradients with `qf1.clone().requires_grad_(False)` would make it slightly more memory efficient I think, but the improvement is probably marginal.

Reply:
That's a good point. Although I would probably keep it as-is because it makes the device `qf1` belongs to more explicit. Also, we were using `torch.no_grad()` so the gradient wouldn't accumulate anyway.

Reply:
Actually, it wasn't `.clone()` but `copy.deepcopy()`. Got bamboozled by my memory somehow.
Let's keep it as it is.